otter/convergence/selfheal.py

```diff
@@ -0,0 +1,169 @@
+"""
+Self heal service
+"""
+
+from effect import Effect
+from effect.do import do, do_return
+
+from kazoo.exceptions import LockTimeout
+from kazoo.protocol.states import KazooState
+
+from toolz.curried import filter
+
+from twisted.application.internet import TimerService
+from twisted.application.service import MultiService
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from txeffect import perform
+
+from otter.convergence.composition import tenant_is_enabled
+from otter.convergence.service import trigger_convergence
+from otter.log.intents import with_log
+from otter.models.intents import GetAllValidGroups, GetScalingGroupInfo
+from otter.models.interface import ScalingGroupStatus
+from otter.util.zk import GetChildren
+
+
+class SelfHeal(MultiService, object):
+    """
+    A service that triggers convergence on all the groups on interval basis.
+    Only one node is allowed to do this.
+    """
+
+    def __init__(self, dispatcher, kz_client, interval, log, clock,
+                 config_func):
+        super(SelfHeal, self).__init__()
+        self.disp = dispatcher
+        self.log = log.bind(otter_service="selfheal")
+        self.lock = kz_client.Lock("/selfheallock")
+        self.kz_client = kz_client
+        self.clock = clock
+        self.time_range = interval - 5
+        self.config_func = config_func
+        self.calls = []
+        timer = TimerService(
+            interval, lambda: self._converge_all().addErrback(self.log.err))
+        timer.clock = clock
+        timer.setServiceParent(self)
+
+    def stopService(self):
+        """
+        Stop service by cancelling any remaining scheduled calls and releasing
+        the lock
+        """
+        super(SelfHeal, self).stopService()
+        self._cancel_scheduled_calls()
+        return self.lock.release()
+
+    def _cancel_scheduled_calls(self):
+        """
+        Cancel any remaining scheduled calls
+        """
+        active = 0
+        for call in self.calls:
+            if call.active():
+                active += 1
+                call.cancel()
+        self.calls = []
+        return active
+
+    def health_check(self):
+        """
+        Return about whether this object has lock
+        """
+        d = is_lock_acquired(self.disp, self.lock)
+        return d.addCallback(lambda b: (True, {"has_lock": b}))
+
+    def _perform(self):
+        d = perform(self.disp, get_groups_to_converge(self.config_func))
+        d.addCallback(self._setup_converges)
```

Comment:
> if the network slows down, might fail to apply backpressure and therefore end up backing up,

Hmm. My understanding is each looping call is independent of other looping calls and slow down of one/some/all of them should not affect each other. I specifically chose the algorithm because of this. 

> This batch-convergence should probably be wrapped in a gatherResults

I don't want to use `gatherResults` as they will all be triggered at same time. I want to space out triggering evenly over the 5 min period. 

I understand the discomfort with extra state. In fact in the beginning I tried to do "get all groups and trigger convergence spaced out" as an effect function which would be called on each interval but it would result in one bad/slow call blocking other calls.